Makefile

```diff
@@ -1,47 +1,69 @@
+#-----------------------------------------
+#	 Setup
+#-----------------------------------------
+
 NAME_SRV = hdb++es-srv
+PREFIX=/usr/local
+
+#-----------------------------------------
+#	 Include Paths
+#-----------------------------------------
 
-LIBHDBPP_DIR = .libhdbpp
-LIBHDBPP_INC = ./$(LIBHDBPP_DIR)/src
-LIBHDBPP_LIB = ./$(LIBHDBPP_DIR)/lib
+ifdef TANGO_INC
+	INC_DIR += -I${TANGO_INC}
+endif
 
-CXXFLAGS += -DRELEASE='"$HeadURL$ "' -I./$(LIBHDBPP_INC)
-LDFLAGS = -lhdb++ -L./$(LIBHDBPP_LIB)
+ifdef TANGO_LIB
+	LIB_DIR	+= -L${TANGO_LIB}
+endif
 
-TANGO_INC := ${TANGO_DIR}/include/tango
-OMNIORB_INC := ${OMNIORB_DIR}/include
-ZMQ_INC :=  ${ZMQ_DIR}/include
+ifdef OMNIORB_INC
+	INC_DIR	+= -L${OMNIORB_INC}
+endif
 
-TANGO_LIB = ${TANGO_DIR}/lib
-OMNIORB_LIB = ${OMNIORB_DIR}/lib
-ZMQ_LIB = ${ZMQ_DIR}/lib
+ifdef OMNIORB_LIB
+	LIB_DIR	+= -L${OMNIORB_LIB}
+endif
 
-INC_DIR = -I${TANGO_INC} -I${OMNIORB_INC} -I${ZMQ_INC}
-LIB_DIR = -L${TANGO_LIB} -L${OMNIORB_LIB} -L${ZMQ_LIB} -L/usr/local/lib
+ifdef ZMQ_INC
+	INC_DIR += -I${ZMQ_INC}
+endif
+
+ifdef LIBHDBPP_INC
+	INC_DIR += -I${LIBHDBPP_INC}
+endif
+
+ifdef LIBHDBPP_LIB
+	LIB_DIR	+= -L${LIBHDBPP_LIB}
+endif
 
 #-----------------------------------------
 #	 Default make entry
 #-----------------------------------------
+
 default: release
 release debug: bin/$(NAME_SRV)
 
 #-----------------------------------------
 #	Set CXXFLAGS and LDFLAGS
 #-----------------------------------------
+
 CXXFLAGS += -std=gnu++0x -D__linux__ -D__OSVERSION__=2 -pedantic -Wall \
-	-Wno-non-virtual-dtor -Wno-long-long -DOMNI_UNLOADABLE_STUBS \
-	$(INC_DIR) -Isrc
+	-Wno-non-virtual-dtor -Wno-long-long -DOMNI_UNLOADABLE_STUBS $(INC_DIR) -Isrc
+
 ifeq ($(GCCMAJOR),4)
     CXXFLAGS += -Wextra
 endif
 ifeq ($(GCCMAJOR),5)
     CXXFLAGS += -Wextra
 endif
-LDFLAGS += $(LIB_DIR) -ltango -llog4tango -lomniORB4 -lomniDynamic4 \
-	-lCOS4 -lomnithread -lzmq
+
+LDFLAGS += $(LIB_DIR) -lhdb++ -ltango -llog4tango -lomniORB4 -lomniDynamic4 -lomnithread
```

Comment:
-lCOS4 has to be specified to compile, and I don't get the reason to remove -lzmq too

Reply:
We do not need to link to them. The code does not use any symbols from them libraries. Tango does, and should link to them.

Reply:
without -lCOS4, and empty LD_LIBRARY_PATH, it cannot compile

Reply:
...
g++ -std=gnu++0x -D__linux__ -D__OSVERSION__=2 -pedantic -Wall -Wno-non-virtual-dtor -Wno-long-long -DOMNI_UNLOADABLE_STUBS -I/usr/local/tango-9.2.5a/include/tango -I/usr/local/omniorb-4.2.1/include -I/usr/local/zeromq-4.0.7/include -I/usr/local/libhdbpp-1.0.0/include -Isrc -O2 -DNDEBUG obj/HdbEventSubscriberStateMachine.o obj/CheckPeriodicThread.o obj/HdbEventSubscriberClass.o obj/StatsThread.o obj/HdbEventSubscriber.o obj/HdbDevice.o obj/ClassFactory.o obj/PollerThread.o obj/PushThread.o obj/main.o obj/SubscribeThread.o -o bin/hdb++es-srv -L/usr/local/tango-9.2.5a/lib -L/usr/local/omniorb-4.2.1/lib -L/usr/local/libhdbpp-1.0.0/lib -lhdb++ -ltango -llog4tango -lomniORB4 -lomniDynamic4 -lomnithread -s
/usr/bin/ld: warning: libCOS4.so.2, needed by /usr/local/tango-9.2.5a/lib/libtango.so, not found (try using -rpath or -rpath-link)
/usr/local/tango-9.2.5a/lib/libtango.so: undefined reference to `CosNotifyChannelAdmin::EventChannelFactory::_narrow(CORBA::Object*)'
...

Reply:
Doubler checking my builds. I have not hit this issue.

Can you run ldd on your libtango.so for me please?

Reply:
ldd /usr/local/tango-9.2.5a/lib/libtango.so
	linux-vdso.so.1 =>  (0x00007ffc55971000)
	liblog4tango.so.5 => /usr/local/tango-9.2.5a/lib/liblog4tango.so.5 (0x00007f066614d000)
	libzmq.so.4 => /usr/local/zeromq-4.0.7/lib/libzmq.so.4 (0x00007f0665efd000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f0665cf9000)
	libomniORB4.so.2 => /usr/local/omniorb-4.2.1/lib/libomniORB4.so.2 (0x00007f0665930000)
	libomniDynamic4.so.2 => /usr/local/omniorb-4.2.1/lib/libomniDynamic4.so.2 (0x00007f0665412000)
	libCOS4.so.2 => /usr/local/omniorb-4.2.1/lib/libCOS4.so.2 (0x00007f0664fe7000)
	libomnithread.so.4 => /usr/local/omniorb-4.2.1/lib/libomnithread.so.4 (0x00007f0664de1000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f0664bc4000)
	libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f0664842000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f0664539000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f066416f000)
	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f0663f58000)
	librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007f0663d50000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f06670ca000)

Reply:
I think the problem there is you have everything installed somewhere ldconfig will not look for them, unless you tell it. The actual issue is a linkage between libtango and libCOS4, not hdbpp-es and libCOS4. The undefined reference is for libtango.

Try export LD_LIBRARY_PATH=/usr/local/omniorb-4.2.1/lib

Reply:
Or try installing the various dependencies under /usr/local/lib /usr/lib etc?

Reply:
I know that exporting  LD_LIBRARY_PATH it compiles, but that's not the solution. It means that symbols are required and the dependency needs to be explicitly declared.

Reply:
The symbols are required by libtango 

From the compile warning: " warning: libCOS4.so.2, needed by /usr/local/tango-9.2.5a/lib/libtango.so" 

From your ldd: "libCOS4.so.2 => /usr/local/omniorb-4.2.1/lib/libCOS4.so.2 (0x00007f0664fe7000)"

The linker can not find them, since you have installed libCOS4 outside of its normal search paths: "/usr/local/omniorb-4.2.1/lib/libCOS4.so.2"

The solution is not to link hdbpp-es against libCOS, its to tell the linker where you have installed the libraries so it can resolve symbols.

The undefined reference is for tango: "/usr/local/tango-9.2.5a/lib/libtango.so: undefined reference to `CosNotifyChannelAdmin::EventChannelFactory::_narrow(CORBA::Object*)'"

Now, if it was for hdbpp-es, then we must link hdbpp-es against it.

Reply:
Or, sorry for got to add, install libCOS4 where the linker searches, for example, /usr/lib

Reply:
$ pkg-config --libs tango
-L/usr/local/tango-9.2.5a/lib -L/usr/local/omniorb-4.2.1/lib -L/usr/local/lib -ltango -lomniDynamic4 -lCOS4 -lomniORB4 -lomnithread -llog4tango -lzmq

Reply:
Ok, just checked, I do not get these problems with everything installed to /usr/lib.

Reply:
We are not using that as part (pk-config) of the build here (in general or for hdb++). But please trust me, the problem is not undefined symbols in hdbpp-es, they are in libtango, see my comment above.  The undefined reference would be in hdbpp-es.

An example, unset the OMNI_LIB export and look at the error.

Reply:
We are adding an extra flag to allow both situations!